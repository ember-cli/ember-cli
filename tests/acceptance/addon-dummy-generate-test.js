'use strict';

const ember = require('../helpers/ember');
const fs = require('fs-extra');
const path = require('path');
let root = process.cwd();
let tmproot = path.join(root, 'tmp');
const Blueprint = require('../../lib/models/blueprint');
const BlueprintNpmTask = require('ember-cli-internal-test-helpers/lib/helpers/disable-npm-on-blueprint');
const mkTmpDirIn = require('../../lib/utilities/mk-tmp-dir-in');

const chai = require('../chai');
let expect = chai.expect;
let file = chai.file;

describe('Acceptance: ember generate in-addon-dummy', function () {
  this.timeout(20000);

  before(function () {
    BlueprintNpmTask.disableNPM(Blueprint);
  });

  after(function () {
    BlueprintNpmTask.restoreNPM(Blueprint);
  });

  beforeEach(async function () {
    let tmpdir = await mkTmpDirIn(tmproot);
    process.chdir(tmpdir);
  });

  afterEach(function () {
    process.chdir(root);
    return fs.remove(tmproot);
  });

  function initAddon() {
    return ember(['addon', 'my-addon', '--skip-npm', '--skip-bower']).then(addJSHint);
  }

  function addJSHint() {
    let pkg = fs.readJsonSync('package.json');
    pkg.devDependencies['ember-cli-jshint'] = '*';
    fs.writeJsonSync('package.json', pkg);
  }

  function generateInAddon(args) {
    let generateArgs = ['generate'].concat(args);

    return initAddon().then(function () {
      return ember(generateArgs);
    });
  }

  it('dummy blueprint foo', async function () {
    await generateInAddon(['blueprint', 'foo', '--dummy']);

    expect(file('blueprints/foo/index.js').content).to.matchSnapshot();
  });

  it('dummy blueprint foo/bar', async function () {
    await generateInAddon(['blueprint', 'foo/bar', '--dummy']);

    expect(file('blueprints/foo/bar/index.js').content).to.matchSnapshot();
  });

  it('dummy http-mock foo', async function () {
    await generateInAddon(['http-mock', 'foo', '--dummy']);

    expect(file('server/index.js').content).to.matchSnapshot();

    expect(file('server/mocks/foo.js').content).to.matchSnapshot();

    expect(file('server/.jshintrc').content).to.matchSnapshot();
  });

  it('dummy http-mock foo-bar', async function () {
    await generateInAddon(['http-mock', 'foo-bar', '--dummy']);

    expect(file('server/index.js').content).to.matchSnapshot();

    expect(file('server/mocks/foo-bar.js').content).to.matchSnapshot();

    expect(file('server/.jshintrc').content).to.matchSnapshot();
  });

  it('dummy http-proxy foo', async function () {
    await generateInAddon(['http-proxy', 'foo', 'http://localhost:5000', '--dummy']);

    expect(file('server/index.js').content).to.matchSnapshot();

    expect(file('server/proxies/foo.js').content).to.matchSnapshot();

    expect(file('server/.jshintrc').content).to.matchSnapshot();
  });

  it('dummy server', async function () {
    await generateInAddon(['server', '--dummy']);
    expect(file('server/index.js')).to.exist;
  });

  // ember addon foo --lang
  // -------------------------------
  // Good: Correct Usage
  it('ember addon foo --lang=(valid code): no message + set `lang` in index.html', async function () {
    await ember(['addon', 'foo', '--skip-npm', '--skip-bower', '--skip-git', '--lang=en-US']);
    expect(file('tests/dummy/app/index.html')).to.contain('<html lang="en-US">');
  });

  // Edge Case: both valid code AND programming language abbreviation, possible misuse
  it('ember addon foo --lang=(valid code + programming language abbreviation): emit warning + set `lang` in index.html', async function () {
    await ember(['addon', 'foo', '--skip-npm', '--skip-bower', '--skip-git', '--lang=css']);
    expect(file('tests/dummy/app/index.html')).to.contain('<html lang="css">');
  });

  // Misuse: possibly an attempt to set app programming language
  it('ember addon foo --lang=(programming language): emit warning + do not set `lang` in index.html', async function () {
    await ember(['addon', 'foo', '--skip-npm', '--skip-bower', '--skip-git', '--lang=JavaScript']);
    expect(file('tests/dummy/app/index.html')).to.contain('<html>');
  });

  // Misuse: possibly an attempt to set app programming language
  it('ember addon foo --lang=(programming language abbreviation): emit warning + do not set `lang` in index.html', async function () {
    await ember(['addon', 'foo', '--skip-npm', '--skip-bower', '--skip-git', '--lang=JS']);
    expect(file('tests/dummy/app/index.html')).to.contain('<html>');
  });

  // Misuse: possibly an attempt to set app programming language
  it('ember addon foo --lang=(programming language file extension): emit warning + do not set `lang` in index.html', async function () {
    await ember(['addon', 'foo', '--skip-npm', '--skip-bower', '--skip-git', '--lang=.js']);
    expect(file('tests/dummy/app/index.html')).to.contain('<html>');
  });

  // Misuse: Invalid Country Code
  it('ember addon foo --lang=(invalid code): emit warning + do not set `lang` in index.html', async function () {
    await ember(['addon', 'foo', '--skip-npm', '--skip-bower', '--skip-git', '--lang=en-UK']);
    expect(file('tests/dummy/app/index.html')).to.contain('<html>');
  });
});
