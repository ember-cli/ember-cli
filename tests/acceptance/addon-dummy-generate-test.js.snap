// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Acceptance: ember generate in-addon-dummy dummy blueprint foo 1`] = `
"'use strict';

module.exports = {
  description: ''

  // locals(options) {
  //   // Return custom template variables here.
  //   return {
  //     foo: options.entity.options.foo
  //   };
  // }

  // afterInstall(options) {
  //   // Perform extra work here.
  // }
};
"
`;

exports[`Acceptance: ember generate in-addon-dummy dummy blueprint foo/bar 1`] = `
"'use strict';

module.exports = {
  description: ''

  // locals(options) {
  //   // Return custom template variables here.
  //   return {
  //     foo: options.entity.options.foo
  //   };
  // }

  // afterInstall(options) {
  //   // Perform extra work here.
  // }
};
"
`;

exports[`Acceptance: ember generate in-addon-dummy dummy http-mock foo 1`] = `
"'use strict';

// To use it create some files under \`mocks/\`
// e.g. \`server/mocks/ember-hamsters.js\`
//
// module.exports = function(app) {
//   app.get('/ember-hamsters', function(req, res) {
//     res.send('hello');
//   });
// };

module.exports = function(app) {
  const globSync   = require('glob').sync;
  const mocks      = globSync('./mocks/**/*.js', { cwd: __dirname }).map(require);
  const proxies    = globSync('./proxies/**/*.js', { cwd: __dirname }).map(require);

  // Log proxy requests
  const morgan = require('morgan');
  app.use(morgan('dev'));

  mocks.forEach(route => route(app));
  proxies.forEach(route => route(app));
};
"
`;

exports[`Acceptance: ember generate in-addon-dummy dummy http-mock foo 2`] = `
"'use strict';

module.exports = function(app) {
  const express = require('express');
  let fooRouter = express.Router();

  fooRouter.get('/', function(req, res) {
    res.send({
      'foo': []
    });
  });

  fooRouter.post('/', function(req, res) {
    res.status(201).end();
  });

  fooRouter.get('/:id', function(req, res) {
    res.send({
      'foo': {
        id: req.params.id
      }
    });
  });

  fooRouter.put('/:id', function(req, res) {
    res.send({
      'foo': {
        id: req.params.id
      }
    });
  });

  fooRouter.delete('/:id', function(req, res) {
    res.status(204).end();
  });

  // The POST and PUT call will not contain a request body
  // because the body-parser is not included by default.
  // To use req.body, run:

  //    npm install --save-dev body-parser

  // After installing, you need to \`use\` the body-parser for
  // this mock uncommenting the following line:
  //
  //app.use('/api/foo', require('body-parser').json());
  app.use('/api/foo', fooRouter);
};
"
`;

exports[`Acceptance: ember generate in-addon-dummy dummy http-mock foo 3`] = `
"{
  \\"node\\": true
}
"
`;

exports[`Acceptance: ember generate in-addon-dummy dummy http-mock foo-bar 1`] = `
"'use strict';

// To use it create some files under \`mocks/\`
// e.g. \`server/mocks/ember-hamsters.js\`
//
// module.exports = function(app) {
//   app.get('/ember-hamsters', function(req, res) {
//     res.send('hello');
//   });
// };

module.exports = function(app) {
  const globSync   = require('glob').sync;
  const mocks      = globSync('./mocks/**/*.js', { cwd: __dirname }).map(require);
  const proxies    = globSync('./proxies/**/*.js', { cwd: __dirname }).map(require);

  // Log proxy requests
  const morgan = require('morgan');
  app.use(morgan('dev'));

  mocks.forEach(route => route(app));
  proxies.forEach(route => route(app));
};
"
`;

exports[`Acceptance: ember generate in-addon-dummy dummy http-mock foo-bar 2`] = `
"'use strict';

module.exports = function(app) {
  const express = require('express');
  let fooBarRouter = express.Router();

  fooBarRouter.get('/', function(req, res) {
    res.send({
      'foo-bar': []
    });
  });

  fooBarRouter.post('/', function(req, res) {
    res.status(201).end();
  });

  fooBarRouter.get('/:id', function(req, res) {
    res.send({
      'foo-bar': {
        id: req.params.id
      }
    });
  });

  fooBarRouter.put('/:id', function(req, res) {
    res.send({
      'foo-bar': {
        id: req.params.id
      }
    });
  });

  fooBarRouter.delete('/:id', function(req, res) {
    res.status(204).end();
  });

  // The POST and PUT call will not contain a request body
  // because the body-parser is not included by default.
  // To use req.body, run:

  //    npm install --save-dev body-parser

  // After installing, you need to \`use\` the body-parser for
  // this mock uncommenting the following line:
  //
  //app.use('/api/foo-bar', require('body-parser').json());
  app.use('/api/foo-bar', fooBarRouter);
};
"
`;

exports[`Acceptance: ember generate in-addon-dummy dummy http-mock foo-bar 3`] = `
"{
  \\"node\\": true
}
"
`;

exports[`Acceptance: ember generate in-addon-dummy dummy http-proxy foo 1`] = `
"'use strict';

// To use it create some files under \`mocks/\`
// e.g. \`server/mocks/ember-hamsters.js\`
//
// module.exports = function(app) {
//   app.get('/ember-hamsters', function(req, res) {
//     res.send('hello');
//   });
// };

module.exports = function(app) {
  const globSync   = require('glob').sync;
  const mocks      = globSync('./mocks/**/*.js', { cwd: __dirname }).map(require);
  const proxies    = globSync('./proxies/**/*.js', { cwd: __dirname }).map(require);

  // Log proxy requests
  const morgan = require('morgan');
  app.use(morgan('dev'));

  mocks.forEach(route => route(app));
  proxies.forEach(route => route(app));
};
"
`;

exports[`Acceptance: ember generate in-addon-dummy dummy http-proxy foo 2`] = `
"'use strict';

const proxyPath = '/foo';

module.exports = function(app) {
  // For options, see:
  // https://github.com/nodejitsu/node-http-proxy
  let proxy = require('http-proxy').createProxyServer({});

  proxy.on('error', function(err, req) {
    console.error(err, req.url);
  });

  app.use(proxyPath, function(req, res, next){
    // include root path in proxied request
    req.url = proxyPath + '/' + req.url;
    proxy.web(req, res, { target: 'http://localhost:5000' });
  });
};
"
`;

exports[`Acceptance: ember generate in-addon-dummy dummy http-proxy foo 3`] = `
"{
  \\"node\\": true
}
"
`;

exports[`Acceptance: ember generate in-addon-dummy dummy http-proxy foo 4`] = `
"'use strict';

// To use it create some files under \`mocks/\`
// e.g. \`server/mocks/ember-hamsters.js\`
//
// module.exports = function(app) {
//   app.get('/ember-hamsters', function(req, res) {
//     res.send('hello');
//   });
// };

module.exports = function(app) {
  const globSync   = require('glob').sync;
  const mocks      = globSync('./mocks/**/*.js', { cwd: __dirname }).map(require);
  const proxies    = globSync('./proxies/**/*.js', { cwd: __dirname }).map(require);

  // Log proxy requests
  const morgan = require('morgan');
  app.use(morgan('dev'));

  mocks.forEach(route => route(app));
  proxies.forEach(route => route(app));
};
"
`;

exports[`Acceptance: ember generate in-addon-dummy dummy http-proxy foo 5`] = `
"'use strict';

const proxyPath = '/foo';

module.exports = function(app) {
  // For options, see:
  // https://github.com/nodejitsu/node-http-proxy
  let proxy = require('http-proxy').createProxyServer({});

  proxy.on('error', function(err, req) {
    console.error(err, req.url);
  });

  app.use(proxyPath, function(req, res, next){
    // include root path in proxied request
    req.url = proxyPath + '/' + req.url;
    proxy.web(req, res, { target: 'http://localhost:5001' });
  });
};
"
`;

exports[`Acceptance: ember generate in-addon-dummy dummy http-proxy foo 6`] = `
"{
  \\"node\\": true
}
"
`;
